File: backend/modules/Services/CromaticaCreativa.Modules.Services.Infrastructure/Persistence/Mappers/ServiceMapper.cs
using CromaticaCreativa.Modules.Services.Domain.Aggregates;
using CromaticaCreativa.Modules.Services.Domain.Enums;
using CromaticaCreativa.Modules.Services.Domain.ValueObjects;
using CromaticaCreativa.Modules.Services.Infrastructure.Persistence.Models;

namespace CromaticaCreativa.Modules.Services.Infrastructure.Persistence.Mappers;

public static class ServiceMapper
{
    public static Service ToDomain(ServiceModel model)
    {
        ArgumentNullException.ThrowIfNull(model);

        var service = Service.Create(
            new ServiceId(model.Id),
            new ServiceName(model.Name),
            model.Description,
            new MediaReference(model.ImageReference),
            new DisplayOrder(model.DisplayOrder));

        switch (model.Status)
        {
            case ServicesPersistenceValues.Active:
                service.Activate();
                break;
            case ServicesPersistenceValues.Inactive:
                service.Deactivate();
                break;
            default:
                throw new InvalidOperationException($"Service '{model.Id}' has unsupported status '{model.Status}'.");
        }

        return service;
    }

    public static ServiceModel ToPersistence(Service service)
    {
        ArgumentNullException.ThrowIfNull(service);

        return new ServiceModel
        {
            Id = service.Id.Value,
            Name = service.Name.Value,
            Description = service.Description,
            ImageReference = service.Image.Value,
            Status = service.Status switch
            {
                ServiceStatus.Active => ServicesPersistenceValues.Active,
                ServiceStatus.Inactive => ServicesPersistenceValues.Inactive,
                _ => throw new InvalidOperationException($"Unsupported Domain service status '{service.Status}'.")
            },
            DisplayOrder = service.Order.Value
        };
    }
}

File: backend/src/modules/Services/Services.Infrastructure/Persistence/Mappers/ServiceMapper.ts
import { ServicePersistenceDto } from '../../../Services.Commons/DTOs/ServicePersistenceDto';
import { Service } from '../../../Services.Domain/Aggregates/Service';
import { ServiceStatus } from '../../../Services.Domain/Enums/ServiceStatus';
import { DisplayOrder } from '../../../Services.Domain/ValueObjects/DisplayOrder';
import { MediaReference } from '../../../Services.Domain/ValueObjects/MediaReference';
import { ServiceId } from '../../../Services.Domain/ValueObjects/ServiceId';
import { ServiceName } from '../../../Services.Domain/ValueObjects/ServiceName';
import { ServicePersistenceModel } from '../Models/ServicePersistenceModel';

export class ServiceMapper {
  public static toDomain(model: ServicePersistenceDto): Service {
    const service = Service.create(new ServiceId(model.id), new ServiceName(model.name), model.description,
      new MediaReference(model.imageReference), new DisplayOrder(model.displayOrder));
    if (model.status === ServiceStatus.Active) service.activate();
    else if (model.status === ServiceStatus.Inactive) service.deactivate();
    else throw new Error(`El servicio '${model.id}' tiene el estado no soportado '${model.status}'.`);
    return service;
  }

  public static toPersistence(service: Service): ServicePersistenceModel {
    return Object.assign(new ServicePersistenceModel(), {
      id: service.id.value,
      name: service.name.value,
      description: service.description,
      imageReference: service.image.value,
      status: service.status,
      displayOrder: service.order.value,
      categories: [],
    });
  }
}
